### PR TITLE
Support grouping folder contents

### LINKED ISSUE #1874: Grouping Folder Contents (Group By)
**Is your feature request related to a problem? Please describe.**
Cannot group folder contents like in File Explorer

**Describe the solution you'd like**
Add a 'Group By' option, allowing folder contents to be automatically grouped.

**Describe alternatives you've considered**
None

**Additional context**
image
image

### PR DESCRIPTION
**Pull request by winston-de**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #1874

**Details of Changes**
Add details of changes here.
- [x] Added prototype implementation
- [x] Fix bug where only 32 items are loaded
- [x] Test virtualization
- [x] Add flyout changing grouping and sorting options
- [x] Save grouping mode in folder settings
- [x] Support all layout modes (currently only GridView is supported)
- [x] Implement sorting for groups
- [x] Implement sorting for group contents
- [x] Improve performance of grouping and sorting
- [ ] Collapsible group headers (?) (requires re-templating the ListView and GridView controls)
- [x] Semantic zoom
- [x] Show item count in header
- [x] Add sub header (eg file extension when sorted by file type, date range when grouped by date modified)
- [x] Make group by date created default for downloads folder

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

**Screenshots (optional)**
image
image
image

https://user-images.githubusercontent.com/59544401/117530681-566d7780-af93-11eb-99c1-206876e96b38.mp4

https://user-images.githubusercontent.com/59544401/117530787-ea3f4380-af93-11eb-8304-70ca85632abe.mp4

### CODE REVIEW

**Comment by yaichenbaum** [comment]
@winston-de Should this be the default for the downloads folder?

**Comment by gave92** [comment]
@winston-de I'd suggest testing this with medium/big folders before going to far.. to my knowledge groping turns off listview virtualization and performance may be too bad

**Comment by winston-de** [comment]
> @winston-de Should this be the default for the downloads folder? And yes, semantic zoom would be really cool!

@yair100 Yeah, that's a good idea. I think a good way to do that would to ship default settings for specific folders in general. 
~~I agree that a semantic zoom is good, however it may be better to do it in another pr because it likely requires heavily modifying the existing layout modes.~~

**Comment by winston-de** [comment]
> @winston-de I'd suggest testing this with medium/big folders before going too far.. to my knowledge groping turns off listview virtualization and performance may be too bad

@gave92 According to the docs, virtualization will still work as long as a virtualizing panel is used for the group containers (which it is). However, I cannot test this at the moment as it appears I have broken something in such a way that only 32 items can be loaded xD. I'll follow up once I fix that.

image

**Comment by winston-de** [comment]
Here's a comparison between the loading speeds of with grouping (in release mode) and v1.5.27 (both with caching disabled):

https://user-images.githubusercontent.com/59544401/116963096-ba730180-ac5c-11eb-9116-766ac9223d3b.mp4


https://user-images.githubusercontent.com/59544401/116963088-b646e400-ac5c-11eb-9ebd-579111fdf190.mp4

It's definitely slower with grouping, but it looks like virtualization is still working. It's probably possible to speed it up with some optimizations.

**Comment by winston-de** [comment]
@yair100 No, it should fall back to the default until there are settings saved for the downloads folder.

**Comment by yaichenbaum** [comment]
The headers seem a little big, what do you think about making them smaller?

File: Files/UserControls/NavigationToolbar.xaml [hunk 1]
```diff
@@ -965,6 +968,135 @@
...
+                    Height="32"
+                    x:Load="{x:Bind IsPageTypeNotHome, Mode=OneWay}"
+                    AccessKey="S"
+                    AutomationProperties.Name="Arrangement options"
+                    Background="Transparent"
+                    Style="{StaticResource ToolBarButtonStyle}"
+                    ToolTipService.ToolTip="Arrangement options">
+                    <FontIcon FontSize="14" Glyph="&#xF168;" />
+                    <AppBarButton.Flyout>
+                        <Flyout>
```
Comment (yaichenbaum) [inline Files/UserControls/NavigationToolbar.xaml:984]: Should the flyout be its own page or user control so that it's easier to reuse in other places?

**Comment by yaichenbaum** [comment]
What about putting all the info on the same row?

**Comment by winston-de** [comment]
@yair100 I had tried that before, it made the count text look out of place right next to the info text. The other issue is that the subtext can be long. I'll try it again though.

**Comment by winston-de** [comment]
@yair100 image

**Comment by yaichenbaum** [comment]
That looks better, I think we can further improve the design by putting the item count in parenthesis just like it's done in File Explorer. Another thing we can do is that instead of showing "Last month 4/1/2021", we can use whichever date format the user has set.

**Comment by winston-de** [comment]
I don't think we should follow Windows Explorer's design of using parenthesis, I think it makes it harder to find information because the labels are not consistently positioned and have the same font style. I also think it doesn't look good when there's more information being shown, for example:
image

**Comment by yaichenbaum** [comment]
@winston-de To solve the issue with there being too many items in parenthesis, how about keeping the item count out like you originally planned? I think that it will look better now that it uses the date format from settings.

**Comment by winston-de** [comment]
What do you think of this?
image

**Comment by yaichenbaum** [comment]
@winston-de In that case, I would replace the file extension with the item count, having it on the far right doesn't seem right.

**Comment by winston-de** [comment]
Like this?
image

**Comment by yaichenbaum** [comment]
@winston-de Yeah, just without the file extension on the right (you can remove that part completely).

File: Files/UserControls/ArrangementOptions.xaml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            Width="1"
+            Margin="16,0,16,0"
+            VerticalAlignment="Stretch"
+            Fill="{ThemeResource MenuFlyoutSeparatorThemeBrush}"
+            RadiusX="1"
+            RadiusY="1" />
+
+        <StackPanel
+            Grid.Column="2"
+            Margin="0,0,0,0"
```
Comment (yaichenbaum) [inline Files/UserControls/ArrangementOptions.xaml:75]: Can this be simplified or removed?

File: Files/Views/LayoutModes/ColumnViewBase.xaml [hunk 3]
```diff
@@ -1161,6 +1162,94 @@
...
             <Setter Property="BorderThickness" Value="0,0,1,0" />
         </Style>
+        <Style TargetType="ListViewHeaderItem">
+            <Setter Property="FontFamily" Value="{ThemeResource ContentControlThemeFontFamily}" />
+            <Setter Property="FontSize" Value="{ThemeResource GridViewHeaderItemThemeFontSize}" />
+            <Setter Property="Background" Value="{ThemeResource GridViewHeaderItemBackground}" />
+            <Setter Property="Margin" Value="0,0,0,4" />
+            <Setter Property="Padding" Value="12,8,12,0" />
+            <Setter Property="HorizontalContentAlignment" Value="Stretch" />
+            <Setter Property="CornerRadius" Value="4" />
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/ColumnViewBase.xaml:1172]: Can this use the ControlCornerRadius resource?

File: Files/Views/LayoutModes/ColumnViewBase.xaml [hunk 4]
```diff
@@ -1171,167 +1260,247 @@
...
-                                Foreground="{x:Bind ((local3:CloudDriveSyncStatusUI)SyncStatusUI).Foreground, Mode=OneWay}"
-                                Glyph="{x:Bind ((local3:CloudDriveSyncStatusUI)SyncStatusUI).Glyph, Mode=OneWay}"
-                                Visibility="{x:Bind ((local3:CloudDriveSyncStatusUI)SyncStatusUI).LoadSyncStatus, Mode=OneWay}" />
-                        </Grid>
-                    </Grid>
-                </DataTemplate>
-            </ListView.ItemTemplate>
-        </ListView>
+                                Background="Transparent"
+                                CornerRadius="4"
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/ColumnViewBase.xaml:1297]:
```suggestion
                                CornerRadius="{StaticResource ControlCornerRadius}"
```

File: Files/Views/LayoutModes/ColumnViewBase.xaml [hunk 5]
```diff
@@ -1171,167 +1260,247 @@
...
+            <SemanticZoom.ZoomedOutView>
+                <ListView
+                    HorizontalAlignment="Stretch"
+                    IsItemClickEnabled="True"
+                    ItemsSource="{x:Bind CollectionViewSource.View.CollectionGroups, Mode=OneWay}"
+                    SelectionMode="None">
+                    <ListView.ItemTemplate>
+                        <DataTemplate x:DataType="ICollectionViewGroup">
+                            <StackPanel
+                                Margin="0,5,0,5"
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/ColumnViewBase.xaml:1474]: Is the odd margin on purpose?

File: Files/Views/LayoutModes/GenericFileBrowser2.xaml [hunk 6]
```diff
@@ -58,6 +59,95 @@
...
             </Style>
+
+            <Style TargetType="ListViewHeaderItem">
+                <Setter Property="FontFamily" Value="{ThemeResource ContentControlThemeFontFamily}" />
+                <Setter Property="FontSize" Value="{ThemeResource GridViewHeaderItemThemeFontSize}" />
+                <Setter Property="Background" Value="{ThemeResource GridViewHeaderItemBackground}" />
+                <Setter Property="Margin" Value="0,0,0,4" />
+                <Setter Property="Padding" Value="12,8,12,0" />
+                <Setter Property="HorizontalContentAlignment" Value="Stretch" />
+                <Setter Property="CornerRadius" Value="4" />
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/GenericFileBrowser2.xaml:70]: Can you use ControlCornerRadius?

File: Files/Views/LayoutModes/GridViewBrowser.xaml [hunk 7]
```diff
@@ -320,6 +321,95 @@
...
         </DataTemplate>
+
+        <Style TargetType="GridViewHeaderItem">
+            <Setter Property="FontFamily" Value="{ThemeResource ContentControlThemeFontFamily}" />
+            <Setter Property="FontSize" Value="{ThemeResource GridViewHeaderItemThemeFontSize}" />
+            <Setter Property="Background" Value="{ThemeResource GridViewHeaderItemBackground}" />
+            <Setter Property="Margin" Value="0,0,0,4" />
+            <Setter Property="Padding" Value="12,8,12,0" />
+            <Setter Property="HorizontalContentAlignment" Value="Stretch" />
+            <Setter Property="CornerRadius" Value="4" />
```
Comment (yaichenbaum) [inline Files/Views/LayoutModes/GridViewBrowser.xaml:332]: Can you use ControlCornerRadius?

File: Files/Views/LayoutModes/ColumnViewBase.xaml (hunk 5 above)
Comment (winston-de) [reply Files/Views/LayoutModes/ColumnViewBase.xaml:1474]: No, just a bad habit xD

Comment (yaichenbaum) [reply Files/Views/LayoutModes/ColumnViewBase.xaml:1474]: There are a couple other values like this as well, can you go through all of the ones in this PR?

File: Files/UserControls/ArrangementOptions.xaml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        <StackPanel
+            Grid.Column="2"
+            Orientation="Vertical"
+            Spacing="8">
+            <muxc:RadioButtons
+                x:Uid="NavToolbarSortByRadioButtons"
+                Grid.Column="2"
+                FontWeight="Bold"
+                Header="Sort by"
```
Comment (yaichenbaum) [inline Files/UserControls/ArrangementOptions.xaml:85]:
```suggestion
                Header="Sort By"
```

File: Files/UserControls/ArrangementOptions.xaml [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    <Grid>
+        <Grid.ColumnDefinitions>
+            <ColumnDefinition Width="*" />
+            <ColumnDefinition Width="Auto" />
+            <ColumnDefinition Width="*" />
+        </Grid.ColumnDefinitions>
+        <muxc:RadioButtons
+            x:Uid="NavToolbarGroupByRadioButtons"
+            FontWeight="Bold"
+            Header="Group by"
```
Comment (yaichenbaum) [inline Files/UserControls/ArrangementOptions.xaml:30]:
```suggestion
            Header="Group By"
```

File: Files/Strings/en-US/Resources.resw [hunk 10]
```diff
@@ -2180,6 +2180,102 @@
...
     <value>Sidebar resizer</value>
+  </data>
+  <data name="NavToolbarArrangementOptionsButton.ToolTipService.ToolTip" xml:space="preserve">
+    <value>Arrangement options</value>
+  </data>
+  <data name="NavToolbarArrangementOptionsButton.AutomationProperties.Name" xml:space="preserve">
+    <value>Arrangement options</value>
+  </data>
+  <data name="NavToolbarGroupByRadioButtons.Header" xml:space="preserve">
+    <value>Group by</value>
```
Comment (yaichenbaum) [inline Files/Strings/en-US/Resources.resw:2191]:
```suggestion
    <value>Group By</value>
```

File: Files/Strings/en-US/Resources.resw [hunk 11]
```diff
@@ -2180,6 +2180,102 @@
...
+    <value>Date deleted</value>
+  </data>
+  <data name="NavToolbarArrangementOption_OriginalFolder.Content" xml:space="preserve">
+    <value>Original folder</value>
+  </data>
+  <data name="NavToolbarArrangementOption_Type.Content" xml:space="preserve">
+    <value>Type</value>
+  </data>
+  <data name="NavToolbarSortByRadioButtons.Header" xml:space="preserve">
+    <value>Sort by</value>
```
Comment (yaichenbaum) [inline Files/Strings/en-US/Resources.resw:2218]:
```suggestion
    <value>Sort By</value>
```

**Comment by winston-de** [comment]
image

**Review by yair100 — APPROVED** [review]
Great work! LGTM

**Comment by Suyash1266** [comment]
I am unable to find this grouping option either in the app or the docs. Is this unavailable in the windows store version or do I need to enable any settings? Please help. TIA

**Comment by winston-de** [comment]
@Suyash1266 We haven't released this feature yet, it will be included in the next update. And unfortunately, due to a store issue, we haven't been able to ship insider builds with this feature either. Thanks for your patience 🙂

**Comment by Suyash1266** [comment]
Ohh got it. Thanks @winston-de 😀

**Comment by IniGisah** [comment]
I'm not sure if here is the right place to report this, but I think I found a bug where "Date Modified" grouping is mixed with last year modified files and folder

I'm using Files - Dev compiled from latest commit (0f53a4241323f64ccaf66d6c0e281a09349e411c)

Screenshot-1159 Files_-_Dev 18-28-52

**Comment by yaichenbaum** [comment]
@IniGisah Can you open a new issue so we can track this?

**Pull request merged**